bpo-29240: PEP 540: Add a new UTF-8 mode

vstinner opened the pull request:
* Add -X utf8 command line option
* Add PYTHONUTF8 environment variable
* Add sys.flags.utf8_mode
* If the LC_CTYPE is "C" at startup: enables automatically the UTF-8
  mode
* In UTF-8 mode, open() now uses UTF-8 encoding by default
* subprocess._args_from_interpreter_flags(): inherit the UTF-8 mode
  using -X utf8 or -X utf8=strict command line options.
* Skip some tests relying on the current locale if the UTF-8 mode is
  enabled.
* Add test_utf8mode.py
* _Py_DecodeUTF8_surrogateescape() gets a new optional parameter to
  return also the length (number of wide characters).
* Update Py_DecodeLocale() and Py_EncodeLocale() to support the UTF-8
  mode

<!-- issue-number: bpo-29240 -->
https://bugs.python.org/issue29240
<!-- /issue-number -->

zooba commented on Doc/library/sys.rst:
```diff
@@ -463,6 +468,10 @@ always available.
       and :func:`_enablelegacywindowsfsencoding` for more information.
 
+   .. versionchanged:: 3.7
+      The UTF-8 mode can now changes the encoding.
```
Either: "UTF-8 mode can now change the encoding" or "UTF-8 mode now changes the encoding", neither of which are very good at explaining what the impact is.

zooba replied:
Is the point that specifying both UTF-8 mode and legacy encoding is not supported? Or specifying UTF-8 mode will override the legacy encoding setting?

zooba commented on Doc/using/cmdline.rst:
```diff
@@ -423,6 +424,9 @@ Miscellaneous options
    .. versionadded:: 3.6
       The ``-X showalloccount`` option.
 
+   .. versionchanged:: 3.7
```
"versionadded"

zooba commented on Include/fileobject.h:
```diff
@@ -28,6 +28,10 @@ PyAPI_DATA(const char *) Py_FileSystemDefaultEncodeErrors;
 PyAPI_DATA(int) Py_HasFileSystemDefaultEncoding;
 
+#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x03070000
+PyAPI_DATA(int) Py_UTF8Mode;
```
Is this an important piece of information? Is it so important that it has to be in the limited API? What would I use this for?

zooba commented on Modules/main.c:
```diff
@@ -383,7 +383,19 @@ Py_Main(int argc, wchar_t **argv)
-            break;
+        }
+        else if (c == 'X') {
+            if (wcscmp(_PyOS_optarg, L"utf8") == 0) {
```
`wcsicmp`? Any need to be case-sensitive here?

zooba reviewed (commented):
Would like to see some improved documentation before this is merged, but I think the rest of the change looks okay.

zooba commented:
Ping @ncoghlan - more changes to startup sequence :)

vstinner commented:
FYI I didn't look again at my code. I only recreated this pull request to not loose the code, since my old PR was on the old CPython repository.

I still have to update the PEP and post it to python-dev.

Questions like "wcsicmp? Any need to be case-sensitive here?" should be asked on the PEP directly, not on the implementation.

You should wait until the PEP is accepted before reviewing the implementation. The PEP may still change.

zooba commented:
@haypo Ah fair enough :) I haven't been paying attention to all the email recently, so I figured it was accepted already.

vstinner committed:
bpo-29240: PEP 540: Add a new UTF-8 mode

* Add -X utf8 command line option, PYTHONUTF8 environment variable
  and a new sys.flags.utf8_mode flag.
* If the LC_CTYPE locale is "C" at startup: enable automatically the
  UTF-8 mode.
* Add _winapi.GetACP(). encodings._alias_mbcs() now calls
  _winapi.GetACP() to get the ANSI code page
* locale.getpreferredencoding() now returns 'UTF-8' in the UTF-8
  mode. As a side effect, open() now uses the UTF-8 encoding by
  default in this mode.
* Py_DecodeLocale() and Py_EncodeLocale() now use the UTF-8 encoding
  in the UTF-8 mode.
* Update subprocess._args_from_interpreter_flags() to handle -X utf8
* Skip some tests relying on the current locale if the UTF-8 mode is
  enabled.
* Add test_utf8mode.py.
* _Py_DecodeUTF8_surrogateescape() gets a new optional parameter to
  return also the length (number of wide characters).
* pymain_get_global_config() and pymain_set_global_config() now
  always copy flag values, rather than only copying if the new value
  is greater than the old value.

methane commented on Programs/python.c:
```diff
@@ -53,9 +60,30 @@ main(int argc, char **argv)
+            Py_UTF8Mode = 1;
+        }
+        setlocale(LC_CTYPE, ctype);
+        PyMem_RawFree(ctype);
```
What is this setlocale() for?

vstinner replied:
The line 64 changes the locale to the user LC_CTYPE locale. This line restores the locale to its previous value.

methane replied:
Shouldn't previous value be got by `setlocale(LC_CTYPE, NULL)` instead of `setlocale(LC_CTYPE, "")`?

vstinner replied:
Oh, you're right. I fixed it in a new commit.

vstinner committed:
Merge branch 'master' into pep540

Fix conflict in Modules/main.c: pymain_set_global_config() and
_Py_CheckHashBasedPycsMode.

vstinner committed:
Fix test_codecs.test_mbcs_alias()

Mock _winapi.GetACP(), not _bootlocale.getpreferredencoding().

vstinner committed:
Fix main(): copy the current LC_CTYPE

methane commented on Programs/python.c:
```diff
@@ -53,9 +60,37 @@ main(int argc, char **argv)
+    /* UTF-8 mode */
+    char *old_ctype = setlocale(LC_CTYPE, NULL);
+    if (old_ctype != NULL) {
+        old_ctype = _PyMem_RawStrdup(old_ctype);
```
Maybe, it can be:

```c
    if (_Py_LegacyLocaleDetected()) {
        Py_UTF8Mode = 1;
        _Py_CoerceLegacyLocale();
    }
```

See here
https://github.com/python/cpython/blob/4ae06c5337e01bdde28bce57b6b9166ad50947e3/Programs/python.c#L75-L77

vstinner replied:
"if (_Py_LegacyLocaleDetected()) { Py_UTF8Mode = 1;"

Right. It works as expected, I updated my PR.

vstinner committed:
Merge branch 'master' into pep540

vstinner committed:
main.c: reuse _Py_LegacyLocaleDetected()
